### DISCUSSION TITLE
Cannot connect to  BasenGreen 10kwH Wall Mounted Battery (TP_BSTBD_23F_304)

### DISCUSSION THREAD

**Discussion by peterhax** [syssi/esphome-tianpower-bms#19] [Q&A] [answered]
Hi there,

I got a problem to connect via ESPHome to my battery with the bluettooth name TP_BSTBD_23F_304.

I am using a C3 super mini Board.

Maybe its not compatible with your code, as the bluettoth name differs from the tested one ... I also checked the discussion where the change to the right MAC was forgotten. [https://github.com/syssi/esphome-tianpower-bms/issues/16](url)

It is the wall mounted version with 51.2V 200AH from basen green. With the mobile app I can connect to the battery and receive information. I changed the MAC Address in the YAML file to 

mac_address: C4:A6:4E:35:C3:3F

the MAC fround with the Basen Green App.

> INFO ESPHome 2024.6.6
> INFO Reading configuration /config/esphome/esphome-web-8cc064.yaml...
> INFO Starting log output from 192.168.178.102 using esphome API
> INFO Successfully connected to tianpower-bms-ble @ 192.168.178.102 in 0.098s
> INFO Successful handshake with tianpower-bms-ble @ 192.168.178.102 in 0.021s
> [12:15:30][I][app:100]: ESPHome version 2024.6.6 compiled on Jul  8 2024, 10:07:41
> [12:15:30][I][app:102]: Project syssi.esphome-tianpower-bms version 1.0.0
> [12:15:31][C][wifi:599]: WiFi:
> [12:15:31][C][wifi:427]:   Local MAC: 64:E8:33:8C:C0:64
> [12:15:31][C][wifi:432]:   SSID: [redacted]
> [12:15:31][C][wifi:435]:   IP Address: 192.168.178.102
> [12:15:31][C][wifi:439]:   BSSID: [redacted]
> [12:15:31][C][wifi:440]:   Hostname: 'tianpower-bms-ble'
> [12:15:31][C][wifi:442]:   Signal strength: -82 dB ▂▄▆█
> [12:15:31][V][wifi:444]:   Priority: -12.0
> [12:15:31][C][wifi:446]:   Channel: 6
> [12:15:31][C][wifi:447]:   Subnet: 255.255.255.0
> [12:15:31][C][wifi:448]:   Gateway: 192.168.178.1
> [12:15:31][C][wifi:449]:   DNS1: 192.168.178.1
> [12:15:31][C][wifi:450]:   DNS2: 0.0.0.0
> [12:15:31][C][logger:185]: Logger:
> [12:15:31][C][logger:186]:   Level: VERY_VERBOSE
> [12:15:31][C][tianpower_bms_ble:547]:   Accuracy Decimals: 3
> [12:15:31][C][esp32_ble:383]: ESP32 BLE:
> [12:15:31][C][esp32_ble:385]:   MAC address: 64:E8:33:8C:C0:66
> [12:15:31][C][esp32_ble:386]:   IO Capability: none
> [12:15:31][C][esp32_ble_tracker:653]: BLE Tracker:
> [12:15:31][C][esp32_ble_tracker:654]:   Scan Duration: 300 s
> [12:15:31][C][esp32_ble_tracker:655]:   Scan Interval: 320.0 ms
> [12:15:31][C][esp32_ble_tracker:656]:   Scan Window: 30.0 ms
> [12:15:31][C][esp32_ble_tracker:657]:   Scan Type: ACTIVE
> [12:15:31][C][esp32_ble_tracker:658]:   Continuous Scanning: True
> [12:15:31][C][ble_client:027]: BLE Client:
> [12:15:31][C][ble_client:028]:   **Address: C4:A6:4E:35:C3:3F**
> [12:15:31][C][ble_client:029]:   Auto-Connect: TRUE
> [12:15:31][C][mdns:115]: mDNS:
> [12:15:31][C][mdns:116]:   Hostname: tianpower-bms-ble
> [12:15:35][W][tianpower_bms_ble:206]: [C4:A6:4E:35:C3:3F] Not connected
> [12:15:45][W][tianpower_bms_ble:206]: [C4:A6:4E:35:C3:3F] Not connected
> [12:15:55][W][tianpower_bms_ble:206]: [C4:A6:4E:35:C3:3F] Not connected

Any suggestions? Thanks for programming this! It is very helpful! :)

**Comment by syssi** [comment]
The `esp32_ble_tracker` should report periodically seen/advertised BLE devices in the air. Could you provide a more complete log including the advertisements / discovered devices? Please make sure there is no other active BLE connection to your BMS. The BMS can handle a single BLE connection only and stops advertising if there is an active connection.

**Comment by peterhax** [comment]
I just made a reboot of the bms and also tried a reset ... no changes in the behaviour ... the MAC seems to be ther same as before the reset (bms power button and after reboot RESET button) and the BASEN App also shows the battery MAC (I did not connect)
[logs_esphome-web-8cc064_initial_run.txt](https://github.com/user-attachments/files/16128210/logs_esphome-web-8cc064_initial_run.txt)
[logs_esphome-web-8cc064_logs_after_BMS_reset_no_other_devices.txt](https://github.com/user-attachments/files/16128211/logs_esphome-web-8cc064_logs_after_BMS_reset_no_other_devices.txt)

Used YAML File:
[basengreen10kwh.txt](https://github.com/user-attachments/files/16128322/basengreen10kwh.txt)


BR and thanks for the fast answer!

**Reply by syssi** [reply]
```
$ cat logs_esphome-web-8cc064_initial_run.txt  | grep Address:
[13:17:09][C][wifi:435]:   IP Address: 192.168.178.102
[13:17:09][C][api:140]:   Address: tianpower-bms-ble.local:6053
[13:33:49][VV][esp32_ble_tracker:448]:   Address: 38:2E:20:7D:3E:C6 (RANDOM)
[13:34:47][VV][esp32_ble_tracker:448]:   Address: 38:2E:20:7D:3E:C6 (RANDOM)
[13:35:02][VV][esp32_ble_tracker:448]:   Address: 38:2E:20:7D:3E:C6 (RANDOM)
[13:35:04][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:07][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:15][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:26][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:26][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:29][VV][esp32_ble_tracker:448]:   Address: 22:8C:6E:23:78:7D (RANDOM)
[13:35:34][VV][esp32_ble_tracker:448]:   Address: 50:2C:96:92:EC:DF (RANDOM)
[13:35:35][VV][esp32_ble_tracker:448]:   Address: 50:2C:96:92:EC:DF (RANDOM)
[13:35:37][VV][esp32_ble_tracker:448]:   Address: 21:B1:E3:AF:8B:70 (RANDOM)
[13:35:39][VV][esp32_ble_tracker:448]:   Address: 60:A3:C4:1C:5E:25 (RANDOM)
[13:35:40][VV][esp32_ble_tracker:448]:   Address: 60:A3:C4:1C:5E:25 (RANDOM)
[13:35:42][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:04][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:04][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:06][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:09][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:12][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:14][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:36:17][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
[13:37:01][VV][esp32_ble_tracker:448]:   Address: 1E:79:4C:EC:AD:A5 (RANDOM)
```

See ESP is unable to pick up the BLE advertisement of your BMS. :-(

Some bluetooth modules advertises a Bluetooth serial connection and a BLE service with different MAC addresses. Don't expect the shown MAC address of the Basen App is correct one. Where can I download the app you are using? Do you have access to another ESP32?

**Comment by peterhax** [comment]
You can download the file here Android APK here:

[https://www.basenpower.com/uploadfile/2023/08/17/20230817164806q9Kby0.apk](https://www.basenpower.com/uploadfile/2023/08/17/20230817164806q9Kby0.apk)

I will check, if I have another ESP32 with bluetooth and WIFI, come back soon ...

**Reply by syssi** [reply]
Could you provide a screenshot of the app?

**Comment by peterhax** [comment]
Sure, here we go...

![Screenshot_20240708-084526low](https://github.com/syssi/esphome-tianpower-bms/assets/127313371/8c6e6d1d-64d1-4221-9914-867a3539ab00)

**Reply by syssi** [reply]
The app picks up all device names starting with `TP` and tries to connect to these services:

```
Service UUID 0000FF00-0000-1000-8000-00805F9B34FB
Notification characteristics: 0000FF01-0000-1000-8000-00805F9B34FB
Command characteristics: 0000FF02-0000-1000-8000-00805F9B34FB
```

This project should be able to establish a connection out of the box.

**Reply by peterhax** [reply]
Okay,
I know have tested it with a ESP32CAM, same behaviour ... wondering which MAC I should enter ... will try to reset BMS and after that the ESP32 later. I will also check if I still can connect with another mobile device to the BMS.

Thanks so far again!

**Reply by syssi** [reply]
I've added a BLE scanner YAML to the project which prints seen Tianpower devices to the log of the ESPHome node. Please give it a try: https://github.com/syssi/esphome-tianpower-bms/pull/20

**Reply by peterhax** [reply]
Hi,

attached is the log file of the BLE tracker

[history.csv](https://github.com/user-attachments/files/16148911/history.csv)

I used this YAML to track ble:

[bletracker.txt](https://github.com/user-attachments/files/16149038/bletracker.txt)

Regarding behaviour with 2 different mobile devices. I can always detect the TP_BSTBD_23F_304 BLE device with MAC C4:A6:4E:35:C3:3F with both devices but only connect with one device at once. If I disconnect one device I can instantly connect with the other device ... Very strange that the BLE scanner does not detect this MAC but I am not to deep onto BLE mechanisms ...

BR :)

**Reply by syssi** [reply]
I suspect your ESP32 models as root cause of the issue. Please try to use a standard ESP32-WROOM-32 and avoid the C3 and S versions just for testing. Another idea is to use the Arduino platform instead of `ESP-IDF`.

**Reply by peterhax** [reply]
Okay, will try to get "normal" ESP32 modules. arduino platform did not change anything unfortunatelly.

**Reply by peterhax** [reply, marked as answer]
Hi Syssi,

I got my ESP32 modules ... and tada ... the BLE connection could be established successfully ... it seems like the ESP32S and ESP32 C3 board bluetooth does behave in another way ...

[logs_esphome-web-c386d0_ESP32-WROOM-32.txt](https://github.com/user-attachments/files/16240273/logs_esphome-web-c386d0_ESP32-WROOM-32.txt)

BR and some coffee for your efforts ... :)

**Reply by syssi** [reply]
Awesome! :-)
